File: code/examples-ts/04_folder_pipeline.ts
/**
 * 04 — Folder pipeline
 *
 * A folder acts as an inbox: any document uploaded to it
 * is automatically processed by all linked sifts.
 *
 * Use case: a shared receipts inbox processed by two sifts simultaneously —
 * one for accounting fields, one for vendor categorisation.
 *
 * Requirements:
 *   npm install @sifter-ai/sdk
 *   npx tsx 04_folder_pipeline.ts
 *   # Sifter server running on localhost:8000 (./run.sh)
 */
import { readdir, readFile } from "node:fs/promises";
import { join } from "node:path";
import { SifterClient } from "@sifter-ai/sdk";

const API_URL = "http://localhost:8000";
const API_KEY = "sk-dev";

async function uploadToFolder(folderId: string, dirPath: string): Promise<void> {
  const filenames = (await readdir(dirPath)).filter(f => !f.startsWith("."));
  for (const filename of filenames) {
    const bytes = await readFile(join(dirPath, filename));
    const form = new FormData();
    form.append("file", new Blob([bytes]), filename);
    const res = await fetch(`${API_URL}/api/folders/${folderId}/documents`, {
      method: "POST",
      headers: { "X-API-Key": API_KEY },
      body: form,
    });
    if (!res.ok) throw new Error(await res.text());
    const doc = await res.json() as { filename?: string; id?: string };
    console.log(`  Uploaded: ${doc.filename ?? doc.id}`);
  }
}

async function waitForSift(client: SifterClient, siftId: string): Promise<void> {
  while (true) {
    const sift = await client.getSift(siftId);
    if (sift.status !== "indexing") {
      if (sift.status === "error") throw new Error("Extraction failed");
      return;
    }
    await new Promise(r => setTimeout(r, 2000));
  }
}

const client = new SifterClient({ apiUrl: API_URL, apiKey: API_KEY });

// Create the inbox folder
const folder = await client.createFolder(
  "Receipts Inbox",
  "Shared inbox for expense receipts",
);

// Sift 1: accounting fields
const siftAccounting = await client.createSift(
  "Receipts — Accounting",
  "Extract: date, merchant name, amount, currency, payment method, vat_amount",
);

// Sift 2: expense category
const siftCategory = await client.createSift(
  "Receipts — Category",
  `
  Categorise the receipt into one of: travel, meals, accommodation,
  software, hardware, office supplies, marketing, other.
  Also extract: merchant name, amount, category, category_reason (one sentence).
  `,
);

// Link both sifts to the folder
await folder.addSift(siftAccounting);
await folder.addSift(siftCategory);

const linkedSifts = await folder.sifts();
console.log(`Folder '${folder.name}' linked to ${linkedSifts.length} sifts`);

// Upload receipts — both sifts process every document automatically
await uploadToFolder(folder.id, "./receipts/");

console.log("\nWaiting for accounting extraction...");
await waitForSift(client, siftAccounting.id);

console.log("Waiting for category extraction...");
await waitForSift(client, siftCategory.id);

// Merge results by document
const accountingRecords = await siftAccounting.records();
const categoryRecords   = await siftCategory.records();

const accounting = new Map(
  accountingRecords.map(r => [r.document_id, r.extracted_data as Record<string, unknown>]),
);
const categories = new Map(
  categoryRecords.map(r => [r.document_id, r.extracted_data as Record<string, unknown>]),
);

console.log("\nMerged results:");
for (const [docId, acc] of accounting) {
  const cat      = categories.get(docId) ?? {};
  const merchant = String(acc["merchant_name"] ?? cat["merchant_name"] ?? "—");
  const amount   = String(acc["amount"] ?? "—");
  const currency = String(acc["currency"] ?? "");
  const category = String(cat["category"] ?? "—");
  console.log(`  ${merchant.padEnd(30)} ${amount.padStart(10)} ${currency.padEnd(4)}  [${category}]`);
}

File: code/examples/python/async/04_folder_pipeline.py
"""
04 — Folder pipeline (async)

Two sifts process the same inbox concurrently using asyncio.gather.

Requirements:
    pip install sifter-ai
    SIFTER_API_KEY env var set
"""
import asyncio
from sifter import AsyncSifter


async def main():
    async with AsyncSifter() as s:
        folder = await s.create_folder("/receipts-inbox")

        sift_accounting, sift_category = await asyncio.gather(
            s.create_sift(
                name="Receipts — Accounting",
                instructions="Extract: date, merchant name, amount, currency, payment method, vat_amount",
            ),
            s.create_sift(
                name="Receipts — Category",
                instructions="These are expense receipts. Categorise each one into: travel, meals, accommodation, software, hardware, office supplies, marketing, or other. Also extract the merchant name, amount, and a one-sentence reason for the category.",
            ),
        )

        await asyncio.gather(
            folder.add_sift(sift_accounting),
            folder.add_sift(sift_category),
        )

        page = await folder.sifts()
        print(f"Folder '{folder.name}' linked to {page.total} sifts")

        folder.on(
            "folder.document.uploaded",
            lambda doc: print(f"  Uploaded: {doc.get('filename', doc.get('id'))}"),
        )
        await folder.upload("../../documents/invoices/", on_conflict="replace")

        print("\nWaiting for both extractions concurrently...")
        await asyncio.gather(sift_accounting.wait(), sift_category.wait())

        accounting_records, category_records = await asyncio.gather(
            sift_accounting.records(),
            sift_category.records(),
        )

        accounting = {r.get("document_id"): r.get("extracted_data", {}) for r in accounting_records}
        categories = {r.get("document_id"): r.get("extracted_data", {}) for r in category_records}

        print("\nMerged results:")
        for doc_id, acc in accounting.items():
            cat      = categories.get(doc_id, {})
            merchant = acc.get("merchant_name", cat.get("merchant_name", "—"))
            amount   = acc.get("amount", "—")
            currency = acc.get("currency", "")
            category = cat.get("category", "—")
            print(f"  {merchant:<30} {amount:>10} {currency:<4}  [{category}]")


asyncio.run(main())
